src/shared/environment/Environment.php

```diff
@@ -4,7 +4,7 @@
 use PharIo\FileSystem\Directory;
 use PharIo\FileSystem\Filename;
 
-abstract class Environment {
+abstract class Environment implements AuthConfig {
```

Comment:
I'm not sure I like that `Environment` now is coupled to Authentication this way.

I guess having an explicit `EnvironmentAuthConfig` class that gets an instance of Environment and calls explicit methods on it to get either the github or gitlab token would be more my thing. What do you think?

Reply:
I'm not against it, but I'm more concern about the link to create between `Environment` and its `AuthConfig` counterpart.

I see two possible implementation:

### 1) Having a `hasGithubToken` and `getGithubToken` (same for Gitlab) in `Environment`

It imply that the `AuthConfig` will have a test like
```php
if ($domain === 'api.github.com' && $this->environment->hasGithubToken()) {
    // ... $this->environment->getGithubToken()
}
// Same goes for Gitlab
```

But if we add another _provider_ (let's say Bitbucket), we have to add 2 methods in `Environment` and 2 more tests (one for the _has_ and another for the _get_) in the `AuthConfig`, so we have to change code in 2 different places and each times it's to almost duplicate the existing code

### 2) Exposition environment variables for `AuthConfig`

This will reduce the amount of code to write in `Environment` and only `AuthConfig` will need to be updated, but in the same it also we lose the control over the data that can be read by other components, and potentially we give access to sensible data (who know what people put in their environment variables)

Reply:
I see your point and agree that Option 1 seems sub-optimal - despite the fact I love explicit APIs, a more generic API to read environment variables seems better suited.

Regarding your security concern: The user is in control of the environment variables and the full process has access to all data. I don't see how phive cause any security issues with that. On the contrary: We're providing a tool to install software, something you have to trust to begin with.

And it's our code to access the variables. I don't see how and where this would be uncontrolled or dangerous?

So, I guess the 2nd option is a go from my side.